### PR TITLE
Check if a tag name to load is a valid digest.

### LINKED ISSUE #12968: Load fails on images saved by digest.
Edited `2015-05-04 19:34:32+00:00`: apply [Issue Report Template](https://github.com/docker/docker/blob/master/CONTRIBUTING.md#reporting-other-issues)

Description of problem:

The save command succeeds when referencing an image by its digest. The load command fails with an "Illegal tag name" error. When using save and load, I expect both commands to succeed or both commands to fail.

`docker version`:
```
Client version: 1.6.0
Client API version: 1.18
Go version (client): go1.4.2
Git commit (client): 4749651
OS/Arch (client): linux/amd64
Server version: 1.6.0
Server API version: 1.18
Go version (server): go1.4.2
Git commit (server): 4749651
OS/Arch (server): linux/amd64
```

`docker info`:
```
Containers: 0
Images: 22
Storage Driver: aufs
 Root Dir: /home/[snipped]/docker/aufs
 Backing Filesystem: extfs
 Dirs: 22
 Dirperm1 Supported: false
Execution Driver: native-0.2
Kernel Version: 3.13.0-46-generic
Operating System: Ubuntu precise (12.04.5 LTS)
CPUs: 4
Total Memory: 7.754 GiB
Name: bench
ID: GGQ2:KGA7:X3IC:BFFJ:B5UM:N6SK:VX3R:ZX24:RZKP:EEPU:L7L7:PMOE
WARNING: No swap limit support
```

`uname -a`:
```
Linux bench 3.13.0-46-generic #79~precise1-Ubuntu SMP Tue Mar 10 20:25:33 UTC 2015 x86_64 x86_64 x86_64 GNU/Linux
```

Environment details (AWS, VirtualBox, physical, etc.):

physical

How reproducible:

always

Steps to Reproduce:

1. sudo docker pull registry:2.0
2. sudo docker run -d -p 5000:5000 registry:2.0
3. sudo docker tag registry:2.0 localhost:5000/registry:2.0
4. sudo docker push localhost:5000/registry:2.0
5. sudo docker pull localhost:5000/registry@sha256:91fab37536df5ce6afbecb0ae83eb498c6ee02be2b4c5223220fd42676850b62
6. sudo docker save -o testcase.tar localhost:5000/registry@sha256:91fab37536df5ce6afbecb0ae83eb498c6ee02be2b4c5223220fd42676850b62
7. sudo docker load -i testcase.tar

Actual Results:

FATA[0000] Error response from daemon: Illegal tag name
(sha256:91fab37536df5ce6afbecb0ae83eb498c6ee02be2b4c5223220fd42676850b62): only [A-Za-z0-9_.-] are allowed, minimum 1, maximum 128 in length 

Expected Results:

Image loads.

### PR DESCRIPTION
**Pull request by calavera**
Fixes #12968.

/cc @stevvooe, @dmcgowan because I don't really know what I'm doing.

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

**Comment by calavera** [comment]
Maybe I should put that check somewhere else?

File: graph/tags/tags.go [hunk 1]
```diff
@@ -2,22 +2,20 @@ package tags
...
-)
-
 // ValidateTagName validates the name of a tag
 func ValidateTagName(name string) error {
 	if name == "" {
 		return fmt.Errorf("tag name can't be empty")
 	}
-	if !validTagName.MatchString(name) {
+
+	if !v2.TagNameRegexp.MatchString(name) && !digest.DigestRegexp.MatchString(name) {
```
Comment (stevvooe) [inline graph/tags/tags.go]: This needs to use the anchored version, `TagNameAnchoredRegexp`.

File: graph/tags.go [hunk 2]
```diff
@@ -387,7 +382,7 @@ func validateDigest(dgst string) error {
 	if dgst == "" {
 		return errors.New("digest can't be empty")
 	}
-	if !validDigest.MatchString(dgst) {
+	if !digest.DigestRegexp.MatchString(dgst) {
```
Comment (stevvooe) [inline graph/tags.go]: Use `digest.ParseDigest` to check this.

File: graph/tags.go [hunk 3]
```diff
@@ -387,7 +382,7 @@ func validateDigest(dgst string) error {
...
 		return fmt.Errorf("illegal digest (%s): must be of the form [a-zA-Z0-9-_+.]+:[a-fA-F0-9]+", dgst)
```
Comment (stevvooe) [inline graph/tags.go]:
```
return fmt.Errorf("illegal digest (%s): must be of the form %s", dgst, digest.DigestRegexp)
```

File: graph/tags/tags.go (hunk 1 above)
Comment (calavera) [inline graph/tags/tags.go]: @stevvooe I wonder if this is the right place to check the digest. This `ValidateTagName` function is used in the builder too. I'm guessing checking that's a valid digest is not correct in that context. I'm going to move the digest check outside of this function.

**Comment by calavera** [comment]
@stevvooe I've changed the implementation to only check the digest when we load an image. I think this is more correct than checking in every case we use the `ValidateTagName` function.

**Comment by tiborvass** [comment]
LGTM

File: graph/tags/tags.go (hunk 1 above)
Comment (stevvooe) [inline graph/tags/tags.go]: @calavera It is probably not. We are working on a typesafe package to handle this better. Let's focus on correctness, for nwo.

**Comment by stevvooe** [comment]
LGTM

**Pull request merged**